### ISSUE TITLE
Investigate setting breakpoints inside threads

### ISSUE THREAD

**Issue by aivarannamaa** [thonny/thonny#1222]
https://docs.python.org/3/library/threading.html#threading.settrace

**Comment by aivarannamaa** [comment]
Currently Thonny only supports stopping the main thread. It should be possible to extend the support to all threads, but I'm afraid it would require serious refactoring and I'm not sure how useful it would be in Thonny (the UI would need to be able to show multiple stacks etc.)

An easier approach would be allowing stepping only inside a single thread (be it main or secondary), eg. the one that gets to a breakpoint first. This would require less changes in the UI and back-end.

**Comment by lurch** [comment]
Multi-threaded code is always a tricky problem - perhaps it should be considered "beyond" Thonny's focus on beginners? (as you mentioned in a recent issue)

**Comment by aivarannamaa** [comment]
@lurch , I agree that it's tricky. I just wanted to write my thoughts down after someone asked me about this possibility.

From the user experience viewpoint, supporting stopping inside a single secondary thread shouldn't make the UI more complex. Therefore I'll keep this task open in case there appears a library or framework which is desired by the beginners and which does its work in a thread for some reason. Or, in case I feel irresistible urge to solve this problem :) (which I currently don't).

**Comment by lurch** [comment]
I guess writing a GUI (e.g. in tkinter) might be the place that most beginners first come across threads? :man_shrugging:

**Comment by aivarannamaa** [comment]
Basic usage of Tkinter doesn't involve secondary threads. Tkinter can reliably run only in the main thread. You only need to create a thread for some long process (eg. file download).

**Comment by nicolasdespres** [comment]
I am working on a framework called [Youpy](https://github.com/nicolasdespres/youpy) that uses thread to run user code. I plan to use Thonny with this framework for teaching Python in my class. Since users write their code in a callback triggered on an event, all the user code is executed in a different thread (one per event).

For me it is critical to be able to have Thonny able to stop on breakpoints set in user thread. Without that, my student will have to fallback to the good old debugging technique of the log file, which is sad considering Thonny's advanced debugging features.

Maybe it would be enough at the beginning for the stack view to show only the thread currently being debugged.

**Comment by aivarannamaa** [comment]
@nicolasdespres, Youpy looks really interesting! I've been thinking about how I would make "Scratch in Python", but I'm afraid I'll never reach actions with this.

I'll make some experiments -- maybe this issue is not so difficult than I have thought.

**Comment by nicolasdespres** [comment]
Thank you for your feedback and considering doing some experiment :)
